File: src/runtime/storage-key-factory.ts
/**
 * @license
 * Copyright (c) 2020 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../platform/assert-web.js';
import {StorageKey} from './storage/storage-key.js';
import {Capabilities, CapabilityRange} from './capabilities.js';
import {ArcId} from './id.js';

export abstract class StorageKeyOptions {
  constructor(
      public readonly arcId: ArcId,
      public readonly schemaHash: string,
      protected readonly schemaName: string = null) {}
  abstract location(): string;
  abstract unique(): string;
}

export class ContainerStorageKeyOptions extends StorageKeyOptions {
  constructor(arcId: ArcId, schemaHash: string, schemaName?: string) {
    super(arcId, schemaHash, schemaName);
  }

  unique(): string { return ''; }
  location(): string { return this.arcId.toString(); }
}

export class BackingStorageKeyOptions extends StorageKeyOptions {
  constructor(arcId: ArcId, schemaHash: string, schemaName?: string) {
    super(arcId, schemaHash, schemaName);
  }
  unique(): string {
    return this.schemaName && this.schemaName.length > 0
        ? this.schemaName : this.schemaHash;
  }
  location(): string {
    return this.unique();
  }
}

export abstract class StorageKeyFactory {
  abstract get protocol();

  abstract create(options: StorageKeyOptions): StorageKey;

  abstract capabilities(): Capabilities;

  // Returns true, if the current storage key class can support the given set of Capabilities.
  supports(capabilities: Capabilities): boolean {
    return this.capabilities().containsAll(capabilities);
  }
}

// An interface for selecting a factory, if more than one are available for the Capabilities.
export interface FactorySelector {
  select(factories: StorageKeyFactory[]): StorageKeyFactory;
}

// An implementation of a FactorySelector choosing a factory with a least
// restrictive max capabilities set.
export class SimpleCapabilitiesSelector implements FactorySelector {
  constructor(private readonly sortedProtocols = ['volatile', 'ramdisk', 'memdb', 'db']) {}

  select(factories: StorageKeyFactory[]): StorageKeyFactory {
    assert(factories.length > 0);
    return factories.reduce((res, factory) => {
      if (res === null) return factory;
      const resIndex = this.sortedProtocols.indexOf(res.protocol);
      if (resIndex === -1 || this.sortedProtocols.indexOf(factory.protocol) < resIndex) {
        return factory;
      }
      return res;
    }, null);
  }
}

File: java/arcs/core/storage/StorageKeyFactory.kt
/*
* Copyright 2020 Google LLC.
*
* This code may only be used under the BSD style license found at
* http://polymer.github.io/LICENSE.txt
*
* Code distributed by Google as part of this project is also subject to an additional IP rights
* grant found at
* http://polymer.github.io/PATENTS.txt
*/

package arcs.core.storage

import arcs.core.common.ArcId
import arcs.core.data.Capabilities
import arcs.core.data.Schema

abstract class StorageKeyFactory(val protocol: String, val capabilities: Capabilities) {

    abstract fun create(options: StorageKeyOptions): StorageKey

    /**
     * Returns true, if the current storage key class supports the given set of [Capabilities].
     */
    fun supports(other: Capabilities): Boolean {
        return capabilities.containsAll(other)
    }

    /**
     * Options passed to registered [StorageKey] constructors.
     * @property arcId An identifier of an Arc requesting the [StorageKey]
     * @property entitySchema A schema of an entities that will be stored
     * @property unique A unique component of the [StorageKey]
     * @property location A memory location of the [StorageKey]
     */
    interface StorageKeyOptions {
        val arcId: ArcId
        val entitySchema: Schema
        val unique: String
        val location: String
    }

    data class ContainerStorageKeyOptions(
        override val arcId: ArcId,
        override val entitySchema: Schema
    ) : StorageKeyOptions {
        override val unique: String = ""
        override val location: String = arcId.toString()
    }

    data class BackingStorageKeyOptions(
        override val arcId: ArcId,
        override val entitySchema: Schema
    ) : StorageKeyOptions {
        override val unique: String = with(entitySchema.name?.name) {
            if (isNullOrEmpty()) entitySchema.hash else this as String
        }
        override val location: String = unique
    }
}
